### DISCUSSION TITLE
Getting promises back from Meteor methods (add Meteor.callAsync?)

### DISCUSSION THREAD

**Discussion by danopia** [meteor/meteor#12068] [Feature Requests]
Hi, I'm working on some client code that calls Meteor methods. I imagine this is a pretty common task. Additionally, in my case the methods are for various server-only tasks (concerning files on the server's disk) and many have no possible client side simulations.

It keeps coming up that `Meteor.call()` does not return a Promise and that the only way to receive a result is via a callback function. I ended up writing a rough wrapper to work around this:

```ts
import { Meteor } from "meteor/meteor";

export function meteorCallAsync<T=unknown>(name: string, ...args: any[]) {
  return new Promise<T>((resolve, reject) =>
    Meteor.call(name, ...args, (error: unknown, result: T) => {
      if (error) return reject(error);
      resolve(result);
    })
  );
}
```

Can we get some native support for this? I believe adding `Meteor.callAsync`/`Meteor.applyAsync` would line up well with the 'try getting away from Fibers' work that the `mongo` package is currently undergoing in `2.7.4-beta.2`.

I found some prior conversation as far back as 2015:

* https://github.com/meteor/meteor/pull/5005
* https://github.com/meteor/meteor-feature-requests/issues/327
* https://github.com/meteor/meteor/commit/0cbd25111d1249a61ca7adce23fad5215408c821 - appears to add `Meteor.server.callAsync` on server only - `Meteor.callAsync` still appears missing

**Comment by radekmie** [comment]
I do agree that it stays in align https://github.com/meteor/meteor/discussions/11505. Actually, it's the only way to do it, as right now both [`call` and `apply` on the server use `.await()`](https://github.com/meteor/meteor/blob/966698a93f56c72bd67b46f315ec2a599a40dd5c/packages/ddp-server/livedata_server.js#L1729-L1754).

**Comment by perbergland** [comment]
There are quite a few isomorphic (those that exist with the same signature on client and server) static methods that should have *Async versions and then go through a deprecation phase of the old method names. Many on Meteor level: "call()" as you mention, probably subscribe(), user(). Definitely the DDPStatic methods (call, subscribe).
Most Accounts package methods including those that are server-side only but have mongodb reads or writes. Etc etc.

**Reply by radekmie** [reply]
While I do agree that there's a lot of them, not all actually have to be asynchronous. For example, `subscribe` should stay synchronous, but it _could_ have a `.readyPromise: Promise<void>` field (or method). Similarly, `.userId()` can stay synchronous, but `.user()` does a database call, therefore it has to be asynchronous.
